xncml/parser.py

```diff
@@ -0,0 +1,584 @@
+            agg = xr.merge(items)
+        case _:
+            raise NotImplementedError
+
+    return target.merge(agg)
+
+
+def read_ds(obj: Netcdf, ncml: Path) -> xr.Dataset:
+    """
+    Return dataset defined in <netcdf> element.
+
+    Parameters
+    ----------
+    obj : Netcdf
+      <netcdf> object description.
+    ncml : Path
+      Path to NcML document, sometimes required to follow relative links.
+
+    Returns
+    -------
+    xr.Dataset
+      Dataset defined at <netcdf>' `location` attribute.
+    """
+    if obj.location:
+        location = obj.location.removeprefix("file:")
+        if not Path(location).is_absolute():
+            location = ncml.parent / location
+        return xr.open_dataset(location, decode_times=False)
+
+
+def read_group(target: xr.Dataset, ref: xr.Dataset, obj: Group | Netcdf) -> xr.Dataset:
+    """
+    Parse <group> items, typically <dimension>, <variable>, <attribute> and <remove> elements.
+
+    Parameters
+    ----------
+    target : xr.Dataset
+      Target dataset to be updated.
+    ref : xr.Dataset
+      Reference dataset used to copy content into `target`.
+    obj : Group | Netcdf
+       <netcdf> object description.
+
+    Returns
+    -------
+    xr.Dataset
+      Dataset holding variables and attributes defined in <netcdf> element.
+    """
+    dims = {}
+    for item in obj.choice:
+        match item:
+            case Dimension():
+                dims[item.name] = read_dimension(item)
+            case Variable():
+                target = read_variable(target, ref, item, dims)
+            case Attribute():
+                read_attribute(target, item, ref)
+            case Remove():
+                target = read_remove(target, item)
+            case EnumTypedef():
+                raise NotImplementedError
+            case Group():
+                target = read_group(target, ref, item)
+            case Aggregation():
+                pass  # <aggregation> elements are parsed in `read_netcdf`
+            case _:
+                raise AttributeError
+
+    return target
+
+
+def read_scan(obj: Aggregation.Scan, ncml: Path) -> [xr.Dataset]:
+    """
+    Return list of datasets defined in <scan> element.
+
+    Parameters
+    ----------
+    obj : Aggregation.Scan instance
+      <scan> object description.
+    ncml : Path
+      Path to NcML document, sometimes required to follow relative links.
+
+    Returns
+    -------
+    list
+      List of datasets found by scan.
+    """
+    import re
+
+    if obj.date_format_mark:
+        raise NotImplementedError
+
+    path = Path(obj.location)
+    if not path.is_absolute():
+        path = ncml.parent / path
+
+    files = path.rglob("*") if obj.subdirs else path.glob("*")
+
+    if obj.reg_exp:
+        pat = re.compile(obj.reg_exp)
+        files = filter(pat.match, map(str, files))
+    elif obj.suffix:
+        pat = "*" + obj.suffix
+        files = [f for f in files if f.match(pat)]
```

Comment:
To have clean code as above, one could do:
```
import glob  # this is in the standard lib

glob.fnmatch.filter(files, pat)
```
:nerd_face:

Reply:
Nice ! Will do.